src/utils/components/DiskSourceFlyoutMenu/DiskSourceFlyoutMenu.tsx

```diff
@@ -0,0 +1,99 @@
+import React, { FC, useRef, useState } from 'react';
+
+import { useKubevirtTranslation } from '@kubevirt-utils/hooks/useKubevirtTranslation';
+import {
+  Menu,
+  MenuContainer,
+  MenuContent,
+  MenuItem,
+  MenuList,
+  MenuToggle,
+} from '@patternfly/react-core';
+
+import { getDiskSourceOptionGroups } from '../DiskModal/components/DiskSourceSelect/utils/utils';
+import { SourceTypes } from '../DiskModal/utils/types';
+
+import FlyoutItem from './FlyoutItem/FlyoutItem';
+
+export type DiskSourceFlyoutMenuProps = {
+  className?: string;
+  isTemplate?: boolean;
+  onSelect: (value: SourceTypes) => void;
+};
+
+const DiskSourceFlyoutMenu: FC<DiskSourceFlyoutMenuProps> = ({
+  className,
+  isTemplate = false,
+  onSelect,
+}) => {
+  const { t } = useKubevirtTranslation();
+  const [isOpen, setIsOpen] = useState<boolean>(false);
+  const menuRef = useRef<HTMLDivElement>(null);
+  const toggleRef = useRef<HTMLButtonElement>(null);
+
+  const options = getDiskSourceOptionGroups(isTemplate);
+
+  const onToggleClick = () => {
+    setIsOpen((open) => !open);
+  };
+
+  const onSelectSource = (value: SourceTypes) => {
+    onSelect(value);
+    setIsOpen(false);
+  };
+
+  const toggle = (
+    <MenuToggle
+      className={className}
+      isExpanded={isOpen}
+      onClick={onToggleClick}
+      ref={toggleRef}
+      variant="primary"
+    >
+      {t('Add disk')}
+    </MenuToggle>
+  );
+
+  const menu = (
+    <Menu containsFlyout ref={menuRef}>
+      <MenuContent>
+        <MenuList>
+          {options?.map(({ description, groupLabel, items }) => {
+            const isFlyout = items?.length > 1;
+            return isFlyout ? (
+              <MenuItem
+                description={description}
+                flyoutMenu={<FlyoutItem items={items} onSelect={onSelectSource} />}
+                itemId={groupLabel}
+              >
+                {groupLabel}
+              </MenuItem>
+            ) : (
+              <MenuItem
+                description={description}
+                itemId={items?.[0]?.id}
+                onClick={() => onSelectSource(items?.[0]?.id)}
+              >
+                {items?.[0]?.label}
+              </MenuItem>
+            );
+          })}
+        </MenuList>
+      </MenuContent>
+    </Menu>
+  );
+
+  return (
```

Comment:
i understand why menu and toggle are here, as nested components, as they are deeply coupled, but still consider exporting them to new files and giving them the right Capital naming.

Reply:
Extracting those to separated components is not working well for the MenuContainer as it expects a ReactElement type which have children:
![Screenshot - 2024-06-30T153533 466](https://github.com/kubevirt-ui/kubevirt-plugin/assets/67270715/834e37a0-0064-4e64-aec9-a644109999c1)